### PR TITLE
Two Domain with Single Istio Ingress Controller

### PR DESCRIPTION
**Pull request by anpanigr**
Create two domains using domainhome-in-image model.
Add istio configuration with default readinessPort. 
Deploy istio gateway and virtual service on each domain namespaces.
Add host information to gateway and virtual service configurations.
Put the namespace.org as host configuration 
Verify domain pods runs in ready state and services are created.
Verify login to WebLogic console on domain1 through istio ingress http  port by passing host information in HTTP header.
Deploy a web application to domain1 through istio ingress http port  using host information in HTTP header. 
Access web application through istio http ingress port using host information in HTTP header.
Repeat the same steps for domain2.

Infra Change
  Now istio will be installed and (un) installed only once during entire run. so the individual test need not (un)install istio
  The new istio configuration (VS/Gateway) template has an explicit host info based on unique domain namespace
  New utility method to  access/deploy application thru REST with optional HTTP host information

Kind Cluster Result
https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/603 

MultiNode Cluster
https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-model-in-image-tests/1095

### CODE REVIEW

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioTwoDomainsInImage.java [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+    assertTrue(deployRes, "Failed to deploy Istio DestinationRule");
+
+    int istioIngressPort = getIstioHttpIngressPort();
+    logger.info("Istio Ingress Port is {0}", istioIngressPort);
+
+    String consoleUrl = "http://" + K8S_NODEPORT_HOST + ":" + istioIngressPort + "/console/login/LoginForm.jsp";
+    boolean checkConsole = 
+         checkAppUsingHostHeader(consoleUrl, domainNamespace1 + ".org");
+    assertTrue(checkConsole, "Failed to access WebLogic console on domain1");
+    logger.info("WebLogic console on domain1 is acceesible");
```
Comment (rjeberhard) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioTwoDomainsInImage.java:285]: accessible

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioDomainInImage.java [hunk 2]
```diff
@@ -216,33 +215,26 @@ public void testIstioDomainHomeInImageUsingWdt() {
...
-    logger.info("Validating WebLogic admin server access by login to console");
-    boolean loginSuccessful = assertDoesNotThrow(() -> {
-      return adminNodePortAccessible(istioIngressPort, ADMIN_USERNAME_DEFAULT, ADMIN_PASSWORD_DEFAULT);
-    }, "Access to admin server node port failed");
-    assertTrue(loginSuccessful, "Console login validation failed");
+    String consoleUrl = "http://" + K8S_NODEPORT_HOST + ":" + istioIngressPort + "/console/login/LoginForm.jsp";
+    boolean checkConsole = 
+         checkAppUsingHostHeader(consoleUrl, domainNamespace + ".org");
+    assertTrue(checkConsole, "Failed to access WebLogic console");
+    logger.info("WebLogic console is acceesible");
```
Comment (vanajamukkara) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioDomainInImage.java:222]: accessible typo

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioDomainInImage.java [hunk 3]
```diff
@@ -216,33 +215,26 @@ public void testIstioDomainHomeInImageUsingWdt() {
...
     ExecResult result = null;
     result = DeployUtil.deployUsingRest(K8S_NODEPORT_HOST, 
         String.valueOf(istioIngressPort),
         ADMIN_USERNAME_DEFAULT, ADMIN_PASSWORD_DEFAULT, 
-        clusterName, archivePath);
+        clusterName, archivePath, domainNamespace + ".org");
     assertNotNull(result, "Application deployment failed");
     logger.info("Application deployment returned {0}", result.toString());
-    assertEquals("202", result.stdout(), "Application deployed successfully");
+    assertEquals("202", result.stdout(), "Deployment does not returns HTTP status code 202");
```
Comment (vanajamukkara) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioDomainInImage.java:232]: return

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioTwoDomainsInImage.java [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+    installAndVerifyOperator(opNamespace, domainNamespace1,domainNamespace2);
+  }
+
+  /**
+   * Create two domains using domainhome-in-image model.
+   * Add istio configuration with default readinessPort. 
+   * Deploy istio gateway and virtual service on each domain namespaces.
+   * Add host information to gateway and virtual service configurations.
+   * Put the namespace.org as host configuration 
+   * Verify domain pods runs in ready state and services are created.
```
Comment (vanajamukkara) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioTwoDomainsInImage.java:138]: Verify server pods are in ready state

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioTwoDomainsInImage.java [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+    logger.info("WebLogic console on domain1 is acceesible");
+    Path archivePath = Paths.get(ITTESTS_DIR, "../src/integration-tests/apps/testwebapp.war");
+    ExecResult result = null;
+    result = DeployUtil.deployUsingRest(K8S_NODEPORT_HOST, 
+        String.valueOf(istioIngressPort),
+        ADMIN_USERNAME_DEFAULT, ADMIN_PASSWORD_DEFAULT, 
+        clusterName, archivePath, domainNamespace1 + ".org");
+    assertNotNull(result, "Application deployment failed on domain1");
+    logger.info("Application deployment on domain1 returned {0}", result.toString());
+    assertEquals("202", result.stdout(), "Deployment does not returns HTTP status code 202");
```
Comment (vanajamukkara) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioTwoDomainsInImage.java:294]: Deployment did not return HTTP status code 202

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioTwoDomainsInImage.java [hunk 6]
```diff
@@ -0,0 +1,369 @@
...
+
+    String url = "http://" + K8S_NODEPORT_HOST + ":" + istioIngressPort + "/testwebapp/index.jsp";
+    logger.info("Application Access URL {0}", url);
+    boolean checkApp = checkAppUsingHostHeader(url, domainNamespace1 + ".org");
+    assertTrue(checkApp, "Failed to access WebLogic application on domain1");
+
+    checkConsole = checkAppUsingHostHeader(consoleUrl, domainNamespace2 + ".org");
+    assertTrue(checkConsole, "Failed to access domain2 WebLogic console");
+    logger.info("WebLogic console on domain2 is acceesible");
```
Comment (vanajamukkara) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioTwoDomainsInImage.java:303]: accessible

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioTwoDomainsInImage.java [hunk 7]
```diff
@@ -0,0 +1,369 @@
...
+    result = DeployUtil.deployUsingRest(K8S_NODEPORT_HOST, 
+        String.valueOf(istioIngressPort),
+        ADMIN_USERNAME_DEFAULT, ADMIN_PASSWORD_DEFAULT, 
+        clusterName, archivePath, domainNamespace2 + ".org");
+    assertNotNull(result, "Application deployment on domain2 failed");
+    logger.info("Application deployment on domain2 returned {0}", result.toString());
+    assertEquals("202", result.stdout(), "Deployment does not returns HTTP status code 202");
```
Comment (vanajamukkara) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioTwoDomainsInImage.java:310]: same as above

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/utils/DeployUtil.java [hunk 8]
```diff
@@ -310,4 +309,46 @@ public static ExecResult deployUsingRest(String host, String port,
...
+    ExecResult result = null;
+    StringBuffer curlString = new StringBuffer("status=$(curl --noproxy '*' ");
+    curlString.append(" --user " + userName + ":" + password);
+    curlString.append(" -w %{http_code} --show-error -o /dev/null ")
+        .append("-H 'host: " + hostHeader  + "' ")
+        .append("-H X-Requested-By:MyClient ")
+        .append("-H Accept:application/json  ")
+        .append("-H Content-Type:multipart/form-data ")
+        .append("-H Prefer:respond-async ")
+        .append("-F \"model={ name: 'testwebapp', targets: [ { identity: [ clusters, '")
```
Comment (vanajamukkara) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/utils/DeployUtil.java:336]: should there be one more argument to the method - app name testwebapp ?

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioTwoDomainsInImage.java (hunk 5 above)
Comment (anpanigr) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioTwoDomainsInImage.java:294]: yes, deployment thru curl command returns 202 ( for a new deployment ) and 400 ( if the app is already deployed)

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioDomainInImage.java (hunk 2 above)
Comment (anpanigr) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioDomainInImage.java:222]: done

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioTwoDomainsInImage.java (hunk 1 above)
Comment (anpanigr) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioTwoDomainsInImage.java:285]: done

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioDomainInImage.java (hunk 2 above)
Comment (anpanigr) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioDomainInImage.java:222]: done

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioDomainInImage.java (hunk 3 above)
Comment (anpanigr) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioDomainInImage.java:232]: done

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioTwoDomainsInImage.java (hunk 6 above)
Comment (anpanigr) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioTwoDomainsInImage.java:303]: done

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioTwoDomainsInImage.java (hunk 7 above)
Comment (anpanigr) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioTwoDomainsInImage.java:310]: done

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioTwoDomainsInImage.java (hunk 4 above)
Comment (anpanigr) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioTwoDomainsInImage.java:138]: modified the description

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/utils/DeployUtil.java (hunk 8 above)
Comment (anpanigr) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/utils/DeployUtil.java]: Added a new parameter for the Application Name

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioTwoDomainsInImage.java [hunk 9]
```diff
@@ -0,0 +1,367 @@
...
+   * Deploy a web application to domain1 through istio ingress http port 
+   * using host information in HTTP header. 
+   * Access web application through istio http ingress port using host 
+   * information in HTTP header.
+   * Repeat the same steps for domain2.
+   */
+  @Test
+  @DisplayName("Two WebLogic domainhome-in-image with single istio ingress")
+  @Slow
+  @MustNotRunInParallel
```
Comment (sankarpn) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioTwoDomainsInImage.java:148]: These tests can be run in parallel, is there a reason why it shouldn't be?

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioTwoDomainsInImage.java [hunk 10]
```diff
@@ -0,0 +1,367 @@
...
+    checkServiceExists(adminServerPodName2, domainNamespace2);
+
+    // check managed server services created
+    for (int i = 1; i <= replicaCount; i++) {
+      logger.info("Check managedserver service {0} is created in namespace {1}",
+          managedServerPrefix1 + i, domainNamespace1);
+      checkServiceExists(managedServerPrefix1 + i, domainNamespace1);
+      logger.info("Check managedserver service {0} is created in namespace {1}",
+          managedServerPrefix2 + i, domainNamespace2);
+      checkServiceExists(managedServerPrefix2 + i, domainNamespace2);
```
Comment (sankarpn) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioTwoDomainsInImage.java:229]: The podReady and serviceExists should be done in this order.
1. check admin service exists
2. check admin pod ready
3. check ms service exists
4. check ms pod ready

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/utils/CommonTestUtils.java [hunk 11]
```diff
@@ -1821,12 +1822,51 @@ public static Path generateFileFromTemplate(
   }
 
   /**
-  * Create a persistent volume.
+   * Check the WebLogic application using host information in the header.
```
Comment (sankarpn) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/utils/CommonTestUtils.java:1825]:
```suggestion
   * Check the application running in WebLogic server using host information in the header.
```

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioTwoDomainsInImage.java [hunk 12]
```diff
@@ -0,0 +1,367 @@
...
+import static oracle.weblogic.kubernetes.utils.IstioUtils.deployIstioDestinationRule;
+import static oracle.weblogic.kubernetes.utils.IstioUtils.getIstioHttpIngressPort;
+import static oracle.weblogic.kubernetes.utils.ThreadSafeLogger.getLogger;
+import static org.awaitility.Awaitility.with;
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+@DisplayName("Test to create WebLogic domain in domainhome-in-image model with istio configuration")
```
[the remark this answers is not in the document]
Comment (anpanigr) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioTwoDomainsInImage.java:69]: modified 
@DisplayName("Test to create two WebLogic domains in domainhome-in-image model with istio configuration")

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioDomainInImage.java [hunk 13]
```diff
@@ -216,33 +213,26 @@ public void testIstioDomainHomeInImageUsingWdt() {
...
-      Thread.sleep(5 * 1000);
-    } catch (InterruptedException ie) {
-      //    
-    }
-    assertEquals(200,
-        assertDoesNotThrow(() -> OracleHttpClient.get(url, true),
-            "Accessing sample application on admin server failed")
-            .statusCode(), "Status code not equals to 200");
+    boolean checkApp = checkAppUsingHostHeader(url, domainNamespace + ".org");
+    assertTrue(checkApp, "Failed to access WebLogic appliation");
```
[the remark this answers is not in the document]
Comment (anpanigr) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioDomainInImage.java:235]: Done

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioDomainInImage.java [hunk 14]
```diff
@@ -216,33 +213,26 @@ public void testIstioDomainHomeInImageUsingWdt() {
...
     ExecResult result = null;
     result = DeployUtil.deployUsingRest(K8S_NODEPORT_HOST, 
         String.valueOf(istioIngressPort),
         ADMIN_USERNAME_DEFAULT, ADMIN_PASSWORD_DEFAULT, 
-        clusterName, archivePath);
+        clusterName, archivePath, domainNamespace + ".org", "testwebapp");
     assertNotNull(result, "Application deployment failed");
     logger.info("Application deployment returned {0}", result.toString());
-    assertEquals("202", result.stdout(), "Application deployed successfully");
+    assertEquals("202", result.stdout(), "Deployment does not return HTTP status code 202");
```
[the remark this answers is not in the document]
Comment (anpanigr) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioDomainInImage.java:230]: done

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioTwoDomainsInImage.java [hunk 15]
```diff
@@ -0,0 +1,367 @@
...
+  @BeforeAll
+  public static void initAll(@Namespaces(3) List<String> namespaces) {
+    logger = getLogger();
+    // create standard, reusable retry/backoff policy
+    withStandardRetryPolicy = with().pollDelay(2, SECONDS)
+        .and().with().pollInterval(10, SECONDS)
+        .atMost(5, MINUTES).await();
+
+    // get a new unique opNamespace
+    logger.info("Creating unique namespace for Operator");
```
[the remark this answers is not in the document]
Comment (anpanigr) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioTwoDomainsInImage.java:104]: done

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioTwoDomainsInImage.java [hunk 16]
```diff
@@ -0,0 +1,367 @@
...
+
+    // Label the domain/operator namespace with istio-injection=enabled
+    Map<String, String> labelMap = new HashMap();
+    labelMap.put("istio-injection", "enabled");
+
+    assertDoesNotThrow(() -> addLabelsToNamespace(domainNamespace1,labelMap));
+    assertDoesNotThrow(() -> addLabelsToNamespace(domainNamespace2,labelMap));
+    assertDoesNotThrow(() -> addLabelsToNamespace(opNamespace,labelMap));
+
+    logger.info("Namespaces [{0}, {1}, {2}] Labeled ",opNamespace, domainNamespace1, domainNamespace2);
```
[the remark this answers is not in the document]
Comment (anpanigr) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioTwoDomainsInImage.java:124]: done

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioTwoDomainsInImage.java (hunk 9 above)
Comment (anpanigr) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioTwoDomainsInImage.java:148]: Remove the tag @MustNotRunInParallel

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioTwoDomainsInImage.java (hunk 10 above)
Comment (anpanigr) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItIstioTwoDomainsInImage.java:229]: done

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/utils/CommonTestUtils.java (hunk 11 above)
Comment (anpanigr) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/utils/CommonTestUtils.java:1825]: done

**Pull request merged**